### PR TITLE
LPS-200759 Update picklists created through Stateflow in batch

### PR DESCRIPTION
**Pull request by G-Sa**
Related to: [LPS-200759](https://liferay.atlassian.net/browse/LPS-200759)

### CODE REVIEW

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java [hunk 1]
```diff
@@ -177,9 +170,23 @@ public void predefineObjectFieldSettings(
...
-			_objectStateFlowLocalService.addObjectStateFlow(
-				newObjectField.getUserId(), newObjectField.getObjectFieldId(),
-				objectStateFlow.getObjectStates());
+			if (oldObjectField != null) {
+				ObjectStateFlow serviceBuilderObjectStateFlow =
+					_objectStateFlowLocalService.
+						fetchObjectFieldObjectStateFlow(
+							oldObjectField.getObjectFieldId());
+
+				_objectStateFlowLocalService.updateObjectStateFlowTransitions(
```
Comment (seltong) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:179]: I believe that would be better if we break this method in two. One for delete and one for add the the transitions. What do you think?

**Review by seltong — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java (hunk 1 above)
Comment (seltong) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:179]: We saw [this examples](https://github.com/liferay-objects/liferay-portal/blob/fd512b62535d52ff29cc169abc988ab09f2cf20f/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java#L157-L184) and decided to follow with the current approach.

**Review by seltong — APPROVED** [review]
LGTM!

**Comment by marcelabc** [comment]
Hey @carolmariaabb 
Could you review this PR, please? :smile:

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java [hunk 2]
```diff
@@ -183,6 +183,48 @@ public ObjectStateFlow updateDefaultObjectStateFlow(
 		return null;
 	}
 
+	public void updateObjectStateFlowTransitions(
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java:186]: @G-Sa please add the `@Override` annotation

Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java:186]: @G-Sa rename it to `updateObjectStateFlowObjectstateTransitions` to follow the same pattern from `deleteObjectStateFlowObjectStateTransitions`

Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java:186]: @G-Sa can we move this method to the `ObjectStateTransitionLocalServiceImpl` to be in the same place as the already existing method `deleteObjectStateFlowObjectStateTransitions`?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java [hunk 3]
```diff
@@ -183,6 +183,48 @@ public ObjectStateFlow updateDefaultObjectStateFlow(
...
+						objectStateTransition.
+							getTargetObjectStateListTypeEntryId(),
+						objectStateFlowId);
+
+				_objectStateTransitionLocalService.addObjectStateTransition(
+					userId, objectStateFlowId,
+					sourceObjectState.getObjectStateId(),
+					targetObjectState.getObjectStateId());
+			}
+		}
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java:225]: @G-Sa this code is the same from [here](https://github.com/liferay-objects/liferay-portal/blob/3de804b5974a700730551b5235514725085686c0/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java#L87-L116). 

Please move it to a common method to avoid duplicate code.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java [hunk 4]
```diff
@@ -177,9 +170,23 @@ public void predefineObjectFieldSettings(
 			ObjectStateFlow objectStateFlow =
 				objectFieldSetting.getObjectStateFlow();
 
-			_objectStateFlowLocalService.addObjectStateFlow(
-				newObjectField.getUserId(), newObjectField.getObjectFieldId(),
-				objectStateFlow.getObjectStates());
+			if (oldObjectField != null) {
+				ObjectStateFlow serviceBuilderObjectStateFlow =
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:174]: @G-Sa rename it to `oldObjectStateFlow` to follow the same pattern from `ObjectField oldObjectField`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java [hunk 5]
```diff
@@ -158,13 +158,6 @@ public void predefineObjectFieldSettings(
 			List<ObjectFieldSetting> objectFieldSettings)
 		throws PortalException {
 
-		if (oldObjectField != null) {
-			_objectStateFlowLocalService.updateDefaultObjectStateFlow(
-				newObjectField, oldObjectField);
-
-			return;
-		}
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:166]: @G-Sa why did you remove this code? The `updateDefaultObjectStateFlow` method is only called here and I'm afraid of "discarding" all of its logic.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java [hunk 6]
```diff
@@ -177,9 +170,23 @@ public void predefineObjectFieldSettings(
...
+					newObjectField.getUserId(),
+					serviceBuilderObjectStateFlow.getObjectStateFlowId(),
+					objectStateFlow.getObjectStates());
+			}
+			else {
+				_objectStateFlowLocalService.addObjectStateFlow(
+					newObjectField.getUserId(),
+					newObjectField.getObjectFieldId(),
+					objectStateFlow.getObjectStates());
+			}
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:189]: @G-Sa could you explain to me this new behavior?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java [hunk 7]
```diff
@@ -177,9 +170,23 @@ public void predefineObjectFieldSettings(
 			ObjectStateFlow objectStateFlow =
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:170]: @G-Sa rename it to `newObjectStateFlow` to follow the same pattern from `ObjectField newObjectField`

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java (hunk 5 above)
Comment (G-Sa) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:166]: @carolmariaabb That code meant that if the object field already existed it would only touch the "default" state flow, but wouldn't update the list in any way, so i moved the validation down below to do the whole update. 

But nice catch on the `updateDefaultObjectStateFlow`, its really missing, i will put it after the update of the whole list

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java (hunk 6 above)
Comment (G-Sa) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:189]: The idea is to keep adding the ObjectStateFlow as usual if we are inside a "post" flow, but if we are in an "put" flow we should also be able to update the transitions of the flow, not only the possible states that that flow has.

The `addObjectStateFlow` creates states and transitions already, so no need to touch the transitions there, but before that we would only create/update the states during a "put" flow, but the transitions would be left as they were, which could lead to situations where new states introduced in a put would transition to nowhere

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java (hunk 2 above)
Comment (G-Sa) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateFlowLocalServiceImpl.java:186]: I don't know if it makes much sense. 
Moving the method there would mean one of two things, either splitting it in two, so we have the StateFlow part in this class and move the StateTransitions to the other, OR moving all logic to the `ObjectStateTransitionLocalServiceImpl` which would imply that the child objects (ObjectStateTransitions) would have power to change their "owner" (ObjectStateFlow) and that's usually a bad design idea.

Also, i made this method in this class because it already has other methods that deal with both flow and transitions (see `addObjectStateFlow`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java [hunk 8]
```diff
@@ -174,12 +167,30 @@ public void predefineObjectFieldSettings(
...
+							oldObjectField.getObjectFieldId());
+
+				_objectStateFlowLocalService.
+					updateObjectStateFlowObjectStateTransitions(
+						newObjectField.getUserId(),
+						oldObjectStateFlow.getObjectStateFlowId(),
+						newObjectStateFlow.getObjectStates());
+
+				_objectStateFlowLocalService.updateDefaultObjectStateFlow(
+					newObjectField, oldObjectField);
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:186]: @G-Sa since we are only calling the `updateDefaultObjectStateFlow` here, can we move this call to be inside the `updateObjectStateFlow` method?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java (hunk 8 above)
Comment (G-Sa) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/field/business/type/PicklistObjectFieldBusinessType.java:186]: Since we would need to bring in the information of `oldObjectField` and `newObjectField` inside the `updateObjectStateFlow` method, it doesn't make much sense to do this merge of functions

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:test:sf

**Comment by G-Sa** [comment]
ci:test:sf

**Comment by G-Sa** [comment]
ci:test:sf

**Comment by G-Sa** [comment]
ci:test:sf

**Comment by G-Sa** [comment]
ci:test:sf

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by G-Sa** [comment]
ci:forward:force
